src/runtime_src/driver/include/ert.h

```diff
@@ -109,6 +109,44 @@ struct ert_start_kernel_cmd {
   uint32_t data[1];          /* count-1 number of words */
 };
 
+/**
+ * struct ert_init_kernel_cmd: ERT initialize kernel command format
+ * this command initializes CUs by writing CU registers. CUs are
+ * represented by cu_mask and extra_cu_masks.
+ *
+ * @state:           [3-0] current state of a command
+ * @extra_cu_masks:  [11-10] extra CU masks in addition to mandatory mask
+ * @count:           [22-12] number of words following header
+ * @opcode:          [27-23] 0, opcode for init_kernel
+ * @type:            [31-27] 0, type of init_kernel
+ *
+ * @cu_mask:         first mandatory CU mask
+ * @data:            count-9 number of words representing interpreted payload
+ *
+ * The packet payload is comprised of reserved id field, 8 reserved fields,
+ * a mandatory CU mask, and extra_cu_masks per header field, followed by a
+ * CU register map of size (count - (9 + extra_cu_masks)) uint32_t words.
+ */
+struct ert_init_kernel_cmd {
```

Comment:
Besides the reserved[8] what is the difference between this command and ert_start_kernel_cmd? How is this command used differently?

Reply:
There are cases that CUs are required to be initialized once and only once before it can be run. This init kernel command is designed for that purpose. The differences between this one and ert_start_kernel cmd are
1) This command just initialize CUs but not trigger them to run. For given CUs, this command copy the regmap to the CUs register map, but not the first 16bytes (first 4 WORDs); while ert_start_kernel_cmd will not only copy the regmap, but set AP_START and wait for AP done, etc.
2) For a given CU, if it has been initilized already,  it won't be intialized again; while ert_start_kernel_cmd can be triggered multiple times.
3) This command will initialize all CUs in cu_mask and cu_ext_mask; while ert_start_kerne_cmd will pick only one free CU from the masks and run it.
4) Some of the reserved field will be replaced by CU timeout value. The CU timeout/reset feature will follow after this one when the hardware (CU) support reset feature.

Reply:
ert_start_kernel_cmd is just a packet, it's behavior is from the opcode assigned to the packet.  In hindsight maybe including _start_ in the type name is bad, e.g. there is no reason the existing packet couldn't be used as you describe with the opcode you define.

The only difference that may warrant the new command is the reserved fields, but I am not sure how they are used, e.g. why the are exposed in the packet itself?

Reply:
some of the reserved fields will be replaced by cu_timeout and cu_reset_timeout.